plugins/inputs/modbus/modbus.go

```diff
@@ -173,552 +167,265 @@ func (m *Modbus) Init() error {
+			handler.Timeout = time.Duration(m.Timeout)
+			handler.BaudRate = m.BaudRate
+			handler.DataBits = m.DataBits
+			handler.Parity = m.Parity
+			handler.StopBits = m.StopBits
+			m.handler = handler
+		case "ASCII":
+			handler := mb.NewASCIIClientHandler(u.Path)
+			handler.Timeout = time.Duration(m.Timeout)
+			handler.BaudRate = m.BaudRate
+			handler.DataBits = m.DataBits
+			handler.Parity = m.Parity
+			handler.StopBits = m.StopBits
+			m.handler = handler
+		default:
 			return fmt.Errorf("invalid protocol '%s' - '%s' ", u.Scheme, m.TransmissionMode)
 		}
 	default:
-		return fmt.Errorf("invalid controller")
+		return fmt.Errorf("invalid controller %q", m.Controller)
 	}
-}
 
-func disconnect(m *Modbus) error {
-	u, err := url.Parse(m.Controller)
-	if err != nil {
-		return err
-	}
+	m.handler.SetSlave(m.SlaveID)
+	m.client = mb.NewClient(m.handler)
+	m.isConnected = false
 
-	switch u.Scheme {
-	case "tcp":
-		m.tcpHandler.Close()
-		return nil
-	case "file":
-		if m.TransmissionMode == "RTU" {
-			m.rtuHandler.Close()
-			return nil
-		} else if m.TransmissionMode == "ASCII" {
-			m.asciiHandler.Close()
-			return nil
-		} else {
-			return fmt.Errorf("invalid protocol '%s' - '%s' ", u.Scheme, m.TransmissionMode)
-		}
-	default:
-		return fmt.Errorf("invalid controller")
-	}
+	return nil
 }
 
-func validateFieldContainers(t []fieldContainer, n string) error {
-	nameEncountered := map[string]bool{}
-	for _, item := range t {
-		//check empty name
-		if item.Name == "" {
-			return fmt.Errorf("empty name in '%s'", n)
-		}
-
-		//search name duplicate
-		canonicalName := item.Measurement + "." + item.Name
-		if nameEncountered[canonicalName] {
-			return fmt.Errorf("name '%s' is duplicated in measurement '%s' '%s' - '%s'", item.Name, item.Measurement, n, item.Name)
-		}
-		nameEncountered[canonicalName] = true
-
-		if n == cInputRegisters || n == cHoldingRegisters {
-			// search byte order
-			switch item.ByteOrder {
-			case "AB", "BA", "ABCD", "CDAB", "BADC", "DCBA", "ABCDEFGH", "HGFEDCBA", "BADCFEHG", "GHEFCDAB":
-				break
-			default:
-				return fmt.Errorf("invalid byte order '%s' in '%s' - '%s'", item.ByteOrder, n, item.Name)
-			}
+// Connect to a MODBUS Slave device via Modbus/[TCP|RTU|ASCII]
+func (m *Modbus) connect() error {
+	err := m.handler.Connect()
+	m.isConnected = err == nil
+	return err
+}
 
-			// search data type
-			switch item.DataType {
-			case "UINT16", "INT16", "UINT32", "INT32", "UINT64", "INT64", "FLOAT32-IEEE", "FLOAT64-IEEE", "FLOAT32", "FIXED", "UFIXED":
-				break
-			default:
-				return fmt.Errorf("invalid data type '%s' in '%s' - '%s'", item.DataType, n, item.Name)
-			}
+func (m *Modbus) disconnect() error {
+	err := m.handler.Close()
+	m.isConnected = false
+	return err
+}
 
-			// check scale
-			if item.Scale == 0.0 {
-				return fmt.Errorf("invalid scale '%f' in '%s' - '%s'", item.Scale, n, item.Name)
-			}
+func (m *Modbus) gatherFields() error {
+	for slaveID, requests := range m.requests {
```

Comment:
remove slaveID
`for _, requests := range m.requests {`

Reply:
Done.